File: extract/brickmesh_extract/teeth.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Sander Striker
"""
Tooth positions and mesh phasing.

Where the teeth sit is derivable from the part geometry: sweep the angle around
the rotation axis, keep the points near the tip radius, and the teeth show up as
peaks. From that follows the phase two gears need in order to interleave rather
than butt heads - and that phase is what fixes the rotation of the axle running
through them.
"""
from __future__ import annotations

import numpy as np

from . import ldraw


def _axis_frame(axis: np.ndarray):
    axis = axis / np.linalg.norm(axis)
    tmp = np.array([1.0, 0, 0]) if abs(axis[0]) < 0.9 else np.array([0, 1.0, 0])
    u = np.cross(axis, tmp); u /= np.linalg.norm(u)
    v = np.cross(axis, u)
    return axis, u, v


def tooth_angles(part: str, axis: np.ndarray, teeth: int,
                 tip_frac: float = 0.93) -> np.ndarray:
    """
    Angular position (degrees) of each tooth in the part's own frame.
    Returns `teeth` angles, ascending.
    """
    g = ldraw.geometry(part)
    n, u, v = _axis_frame(axis)
    p = g.verts
    r = np.hypot(p @ u, p @ v)
    tip = r.max()
    sel = r > tip_frac * tip
    if sel.sum() < teeth * 4:
        sel = r > 0.85 * tip
    ang = np.degrees(np.arctan2(p[sel] @ v, p[sel] @ u)) % 360.0

    pitch = 360.0 / teeth
    # Fold every tooth onto one pitch window and take the circular mean.
    # That is far more robust than peak picking on a sparse vertex cloud.
    phase = (ang % pitch) / pitch * 2 * np.pi
    mean = np.arctan2(np.sin(phase).mean(), np.cos(phase).mean())
    offset = (mean / (2 * np.pi)) * pitch % pitch
    return (offset + np.arange(teeth) * pitch) % 360.0


def tooth_sharpness(part: str, axis: np.ndarray, teeth: int) -> float:
    """
    How cleanly the teeth cluster (0..1). Low values mean the extraction is
    unreliable and the resulting phase should not be trusted.
    """
    g = ldraw.geometry(part)
    n, u, v = _axis_frame(axis)
    p = g.verts
    r = np.hypot(p @ u, p @ v)
    sel = r > 0.93 * r.max()
    ang = np.degrees(np.arctan2(p[sel] @ v, p[sel] @ u)) % 360.0
    pitch = 360.0 / teeth
    phase = (ang % pitch) / pitch * 2 * np.pi
    return float(np.hypot(np.sin(phase).mean(), np.cos(phase).mean()))


def axle_symmetry_ok(teeth: int) -> bool:
    """
    A cross axle has 4-fold symmetry, so a gear can be seated on it in four
    ways. If the tooth count is a multiple of 4 those four seatings map teeth
    onto teeth and the seating does not affect phase at all.
    """
    return teeth % 4 == 0


def mesh_phase(part_a: str, teeth_a: int, axis_a: np.ndarray,
               part_b: str, teeth_b: int, axis_b: np.ndarray,
               direction_ab: np.ndarray) -> dict:
    """
    Rotation (deg, about each gear's own axis) so that a tooth of A points at B
    while a gap of B points back at A.
    """
    _, ua, va = _axis_frame(axis_a)
    _, ub, vb = _axis_frame(axis_b)

    d = direction_ab / np.linalg.norm(direction_ab)
    beta_a = np.degrees(np.arctan2(d @ va, d @ ua)) % 360.0
    beta_b = np.degrees(np.arctan2(-d @ vb, -d @ ub)) % 360.0

    ta = tooth_angles(part_a, axis_a, teeth_a)
    tb = tooth_angles(part_b, axis_b, teeth_b)
    pitch_a, pitch_b = 360.0 / teeth_a, 360.0 / teeth_b

    # A: bring a tooth onto the line of centers.
    rot_a = (beta_a - ta[0]) % pitch_a
    # B: bring a GAP onto the line of centers, i.e. a tooth half a pitch away.
    rot_b = (beta_b + pitch_b / 2 - tb[0]) % pitch_b

    return {
        "rot_a_deg": rot_a,
        "rot_b_deg": rot_b,
        "pitch_a_deg": pitch_a,
        "pitch_b_deg": pitch_b,
        "sharpness_a": tooth_sharpness(part_a, axis_a, teeth_a),
        "sharpness_b": tooth_sharpness(part_b, axis_b, teeth_b),
        "axle_seating_free_a": axle_symmetry_ok(teeth_a),
        "axle_seating_free_b": axle_symmetry_ok(teeth_b),
    }


if __name__ == "__main__":
    Z = np.array([0.0, 0, 1])
    print("Tooth positions derived from the geometry\n")
    print(f"  {'part':10s}{'teeth':>7}{'pitch':>8}{'1st tooth':>10}{'sharpness':>10}")
    for p, t in [("3647.dat", 8), ("32270.dat", 12), ("4019.dat", 16),
                 ("32269.dat", 20), ("3648b.dat", 24), ("32498.dat", 36)]:
        ang = tooth_angles(p, Z, t)
        sh = tooth_sharpness(p, Z, t)
        print(f"  {p:10s}{t:7d}{360/t:8.2f}{ang[0]:10.2f}{sh:10.3f}"
              f"   {'axle seating free' if axle_symmetry_ok(t) else 'NOTE: seating fixes phase'}")

File: internal/teeth/teeth.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Sander Striker

// Package teeth finds where a gear's teeth sit, and the phase two of them need
// to interleave.
//
// It is derivable from the part itself: sweep the angle around the rotation
// axis, keep the points out near the tip radius, and the teeth show up as
// peaks. From that follows how far each gear has to be turned so a tooth of one
// meets a gap of the other rather than its tooth — and that phase is what fixes
// the rotation of the axle running through them.
//
// Without it a model has its gears at the right centers with their teeth passing
// through each other, which is right where it is measured and wrong wherever it
// is looked at.
package teeth

import (
	"fmt"
	"math"

	"brickmesh/internal/geom"
	"brickmesh/internal/ldraw"
)

// TrustThreshold is the sharpness below which a phase should not be relied on.
//
// The standard gears read between 0.51 and 0.90 except the 40t, which reads
// 0.26: its teeth are small against its radius, so the rim band picks up more
// that is not a tooth. Both implementations agree on that number, so it is the
// gear and not the reading.
const TrustThreshold = 0.45

// Frame builds two directions across an axis, so a part's points can be read as
// an angle around it.
func Frame(axis geom.Vec3) (u, v geom.Vec3) {
	a := axis.Unit()
	tmp := geom.Vec3{X: 1}
	if math.Abs(a.X) >= 0.9 {
		tmp = geom.Vec3{Y: 1}
	}
	u = a.Cross(tmp).Unit()
	return u, a.Cross(u)
}

// Angles is the angular position of each tooth, in degrees, in the part's own
// frame. It returns one per tooth, ascending.
func Angles(lib *ldraw.Library, part string, axis geom.Vec3, teeth int) ([]float64, error) {
	offset, _, err := phaseOf(lib, part, axis, teeth)
	if err != nil {
		return nil, err
	}
	pitch := 360.0 / float64(teeth)
	out := make([]float64, teeth)
	for i := range out {
		out[i] = math.Mod(offset+float64(i)*pitch, 360)
	}
	return out, nil
}

// Sharpness is how cleanly the teeth cluster, from 0 to 1. A low value means
// the reading is unreliable and the phase it implies should not be trusted.
func Sharpness(lib *ldraw.Library, part string, axis geom.Vec3, teeth int) (float64, error) {
	_, sharp, err := phaseOf(lib, part, axis, teeth)
	return sharp, err
}

// phaseOf folds every tooth onto one pitch window and takes the circular mean,
// which is far steadier than picking peaks out of a sparse cloud of vertices.
func phaseOf(lib *ldraw.Library, part string, axis geom.Vec3, teeth int) (offset, sharpness float64, err error) {
	if teeth <= 0 {
		return 0, 0, fmt.Errorf("%s: a gear needs teeth", part)
	}
	g, err := lib.Geometry(part)
	if err != nil {
		return 0, 0, err
	}
	u, v := Frame(axis)

	// The teeth are the points furthest from the axis.
	tip := 0.0
	radii := make([]float64, len(g.Verts))
	for i, p := range g.Verts {
		radii[i] = math.Hypot(p.Dot(u), p.Dot(v))
		if radii[i] > tip {
			tip = radii[i]
		}
	}
	if tip == 0 {
		return 0, 0, fmt.Errorf("%s: no geometry away from the axis", part)
	}

	angles := tipAngles(g, radii, u, v, tip, 0.93)
	if len(angles) < teeth*4 {
		// Too sparse a rim to read: take a wider band rather than guess from
		// a handful of points.
		angles = tipAngles(g, radii, u, v, tip, 0.85)
	}
	if len(angles) == 0 {
		return 0, 0, fmt.Errorf("%s: no rim points to read the teeth from", part)
	}

	pitch := 360.0 / float64(teeth)
	var sumSin, sumCos float64
	for _, a := range angles {
		phase := math.Mod(a, pitch) / pitch * 2 * math.Pi
		sumSin += math.Sin(phase)
		sumCos += math.Cos(phase)
	}
	n := float64(len(angles))
	mean := math.Atan2(sumSin/n, sumCos/n)
	offset = math.Mod(mean/(2*math.Pi)*pitch, pitch)
	if offset < 0 {
		offset += pitch
	}
	return offset, math.Hypot(sumSin/n, sumCos/n), nil
}

func tipAngles(g *ldraw.Geometry, radii []float64, u, v geom.Vec3, tip, frac float64) []float64 {
	var out []float64
	for i, p := range g.Verts {
		if radii[i] <= frac*tip {
			continue
		}
		a := math.Mod(math.Atan2(p.Dot(v), p.Dot(u))*180/math.Pi, 360)
		if a < 0 {
			a += 360
		}
		out = append(out, a)
	}
	return out
}

// SeatingIsFree reports whether the four ways a gear can sit on a cross axle
// all put teeth where teeth were.
//
// A cross axle has four-fold symmetry, so a gear can be seated four ways. Where
// the tooth count is a multiple of four those seatings map teeth onto teeth and
// the choice does not affect the phase at all.
func SeatingIsFree(teeth int) bool { return teeth%4 == 0 }

// Phase is how far each of two meshing gears must turn about its own axis.
type Phase struct {
	RotA, RotB           float64 // degrees
	PitchA, PitchB       float64
	SharpA, SharpB       float64
	SeatFreeA, SeatFreeB bool
}

// MeshPhase works out the rotation that puts a tooth of A on the line of
// centers and a gap of B facing back at it.
func MeshPhase(lib *ldraw.Library, partA string, teethA int, axisA geom.Vec3,
	partB string, teethB int, axisB geom.Vec3, towardB geom.Vec3) (Phase, error) {

	uA, vA := Frame(axisA)
	uB, vB := Frame(axisB)
	d := towardB.Unit()

	betaA := degrees(math.Atan2(d.Dot(vA), d.Dot(uA)))
	betaB := degrees(math.Atan2(-d.Dot(vB), -d.Dot(uB)))

	anglesA, err := Angles(lib, partA, axisA, teethA)
	if err != nil {
		return Phase{}, err
	}
	anglesB, err := Angles(lib, partB, axisB, teethB)
	if err != nil {
		return Phase{}, err
	}
	sharpA, err := Sharpness(lib, partA, axisA, teethA)
	if err != nil {
		return Phase{}, err
	}
	sharpB, err := Sharpness(lib, partB, axisB, teethB)
	if err != nil {
		return Phase{}, err
	}

	pitchA, pitchB := 360.0/float64(teethA), 360.0/float64(teethB)
	return Phase{
		// A brings a tooth onto the line of centers.
		RotA: wrap(betaA-anglesA[0], pitchA),
		// B brings a GAP onto it, which is a tooth half a pitch away.
		RotB:      wrap(betaB+pitchB/2-anglesB[0], pitchB),
		PitchA:    pitchA,
		PitchB:    pitchB,
		SharpA:    sharpA,
		SharpB:    sharpB,
		SeatFreeA: SeatingIsFree(teethA),
		SeatFreeB: SeatingIsFree(teethB),
	}, nil
}

func degrees(rad float64) float64 { return rad * 180 / math.Pi }

func wrap(v, period float64) float64 {
	r := math.Mod(v, period)
	if r < 0 {
		r += period
	}
	return r
}
